libportal/notification.c

```diff
@@ -129,35 +272,30 @@ xdp_portal_add_notification (XdpPortal *portal,
                              GAsyncReadyCallback callback,
                              gpointer data)
 {
-  GAsyncReadyCallback call_done_cb = NULL;
-  CallDoneData *call_done_data = NULL;
+  GVariantBuilder *builder;
+  GVariantIter *iter;
+  g_autoptr(GTask) task = NULL;
 
   g_return_if_fail (XDP_IS_PORTAL (portal));
   g_return_if_fail (flags == XDP_NOTIFICATION_FLAG_NONE);
 
   ensure_action_invoked_connection (portal);
 
-  if (callback)
-    {
-      call_done_cb = call_done; 
-      call_done_data = g_new (CallDoneData, 1);
-      call_done_data->portal = g_object_ref (portal);
-      call_done_data->callback = callback,
-      call_done_data->data = data;
-    }
+  task = g_task_new (portal, cancellable, callback, data);
+  iter = g_variant_iter_new (notification);
+  builder = g_variant_builder_new (G_VARIANT_TYPE ("(sa{sv})"));
 
-  g_dbus_connection_call (portal->bus,
-                          PORTAL_BUS_NAME,
-                          PORTAL_OBJECT_PATH,
-                          "org.freedesktop.portal.Notification",
-                          "AddNotification",
-                          g_variant_new ("(s@a{sv})", id, notification),
-                          NULL,
-                          G_DBUS_CALL_FLAGS_NONE,
-                          -1,
-                          cancellable,
-                          call_done_cb,
-                          call_done_data);
+  g_variant_builder_add (builder, "s", id);
+  /* This will be closed once the async parsing is done */
+  g_variant_builder_open (builder, G_VARIANT_TYPE ("a{sv}"));
+
+  g_task_set_source_tag (task, xdp_portal_add_notification);
+  g_task_set_task_data (task, call_data_new (iter, builder), call_data_free);
+
+  if (portal->notification_interface_version < 1)
+    get_notification_interface_properties (task);
+  else
+    parse_notification (task);
```

Comment:
You could change the `get_notification_interface_properties` to be async which can finish immediately in some situations and when the dbus call is done otherwise. This allows you to have a single flow:
```c
xdp_portal_add_notification()
{
  get_supported(..., get_supported_cb);
}

get_supported_cb()
{
  finish_get_supported();
  parse_notification();
}
```

Reply:
I don't understand what you suggest

Reply:
The flows through the code are a bit messy and could be organized more linearly by implementing a bunch of tasks and then running one after the other, unconditionally:

1. get supported features
2. read files
3. add the notification

Then handle details in the tasks, for example "get supported features" would be something like this:

```c
static void
get_properties_cb (GObject      *source_object,
                   GAsyncResult *result,
                   gpointer      user_data)
{
  g_autoptr(GTask) task = G_TASK (user_data);
  XdpPortal *portal = g_task_get_source_object (task);
  g_autoptr(GError) error = NULL;
  g_autoptr(GVariant) ret = NULL;
  g_autoptr(GVariant) vardict = NULL;

  ret = g_dbus_connection_call_finish (G_DBUS_CONNECTION (source_object), result, &error);
  if (!ret)
    {
      g_task_return_error (task, g_steal_pointer (&error));
      return;
    }

  g_clear_pointer (&portal->supported_notification_options, g_variant_unref);

  g_variant_get (ret, "(&a{sv})", &vardict);
  if (!g_variant_lookup (vardict, "version", "u", &portal->notification_interface_version))
    portal->notification_interface_version = 1;
  if (!g_variant_lookup (vardict, "SupportedOptions", "v", &portal->supported_notification_options))
    {
      g_auto(GVariantBuilder) builder;
      
      g_variant_builder_init (&builder, G_VARIANT_TYPE_VARDICT);
      portal->supported_notification_options = g_variant_ref_sink (g_variant_builder_end (&builder));
    }

  g_task_return_boolean (task, TRUE);
}

void
get_supported_features (XdpPortal *portal,
                        GCancellable *cancellable,
                        GAsyncReadyCallback callback,
                        gpointer data)
{
  g_autoptr(GTask) task = NULL;

  task = g_task_new (portal, cancellable, callback, data);
  g_task_set_source_tag (task, get_supported_features);

  if (portal->notification_interface_version != 0)
    {
      g_task_return_boolean (task, TRUE);
      return;
    }

  g_dbus_connection_call (portal->bus,
                          PORTAL_BUS_NAME,
                          PORTAL_OBJECT_PATH,
                          "org.freedesktop.DBus.Properties",
                          "GetAll",
                          g_variant_new ("(s)", "org.freedesktop.portal.Notification"),
                          G_VARIANT_TYPE ("(a{sv})"),
                          G_DBUS_CALL_FLAGS_NONE,
                          -1,
                          cancellable,
                          get_properties_cb,
                          g_object_ref (task));
}

gboolean
get_supported_features_finish (XdpPortal     *portal,
                               GAsyncResult  *result,
                               GError       **error)
{
  g_autoptr(GVariant) res = NULL;
  g_return_val_if_fail (XDP_IS_PORTAL (portal), FALSE);
  g_return_val_if_fail (g_task_is_valid (result, portal), FALSE);
  g_return_val_if_fail (g_task_get_source_tag (G_TASK (result)) == get_supported_features, FALSE);

  return g_task_propagate_boolean (G_TASK (result), error);
}
```

So, in other words, embrace the GTask.

Reply:
ah ok, so add additional `GTask` instead of using a single one. i can do that.

Reply:
hmm, you marked `get_supported_features()` as public. Is it just an oversight?

Reply:
Yes, I didn't even compile the code

Reply:
reworked the code to use multiple `GTask` and to have a single code path.